In [19]:
# Chap 15.riss.kr 에서 특정 키워드로 논문 / 학술 자료 검색하기

#Step 1. 필요한 모듈을 로딩합니다
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.service import Service
import time


#Step 2. 사용자에게 검색 관련 정보들을 입력 받습니다.
#print("=" *100)
#print(" 이 크롤러는 서울시민 사이트의 논문 및 학술자료 수집용 웹크롤러입니다.")
#print("=" *100)
#query_txt = input('1.수집할 자료의 키워드는 무엇입니까?: ')


#Step 3. 수집된 데이터를 저장할 파일 이름 입력 받기
f_dir = input("2.파일을 저장할 폴더명만 쓰세요(기본값:c:\\py_temp\\):")
if f_dir == '' :
    f_dir="c:\\py_temp\\"


#Step 4. 크롬 드라이버 설정 및 웹 페이지 열기
s = Service("c:/py_temp/chromedriver.exe")
driver = webdriver.Chrome(service=s)

url = 'https://eungdapso.seoul.go.kr/main.do'
driver.get(url)
driver.maximize_window()
time.sleep(2)


In [20]:

#Step 5. 자동으로 검색어 입력 후 조회하기
#element = driver.find_element(By.ID,'query')
#element.send_keys(query_txt)
#element.send_keys("\n")


#link = soup.find('div','my_current_count_box').find('a')['href']

#//*[@id="content"]/div[2]/a/div

In [21]:

#Step 6.학위 논문 선택하기
#driver.find_element(By.LINK_TEXT,'학위논문').click()
#time.sleep(2)



#driver.find_element(By.CLASS_NAME,"mv_banner_box").click()
#time.sleep(2)

#driver.find_element(By.XPATH,'//*[@id="content"]/div[2]/a/div').click()
#time.sleep(2)



#driver.find_element(By.XPATH,'//*[@id="content"]/div[2]/a').click()
#time.sleep(2)


#driver.find_element(By.XPATH,'//*[@id="content"]/div[2]/a/div').click()
#time.sleep(2)


driver.find_element(By.XPATH,'//*[@id="content"]/div[2]/a/div/div').click()
time.sleep(2)

#//*[@id="content"]/div[2]/a
#//*[@id="content"]/div[2]/a/div
#//*[@id="content"]/div[2]/a/div/div
#//*[@id="content"]/div[2]/a/div/div/div[1]


In [22]:
""""
Selenium의 .click()은 자식 요소를 클릭해도 부모 링크가 실행되는 경우가 많다.

현재 구조

<a>
   div
      div

이런 구조에서는 브라우저가 클릭 이벤트를 부모 a 태그로 전달한다.

이것을 이벤트 버블링(Event Bubbling) 이라고 한다.

즉 실제 동작은 이렇게 된다.

div 클릭
 → a 태그로 이벤트 전달
 → 링크 실행

그래서 다음이 모두 같은 결과를 만든다.

a 클릭
div 클릭
div 안의 div 클릭
"""

'"\nSelenium의 .click()은 자식 요소를 클릭해도 부모 링크가 실행되는 경우가 많다.\n\n현재 구조\n\n<a>\n   div\n      div\n\n이런 구조에서는 브라우저가 클릭 이벤트를 부모 a 태그로 전달한다.\n\n이것을 이벤트 버블링(Event Bubbling) 이라고 한다.\n\n즉 실제 동작은 이렇게 된다.\n\ndiv 클릭\n → a 태그로 이벤트 전달\n → 링크 실행\n\n그래서 다음이 모두 같은 결과를 만든다.\n\na 클릭\ndiv 클릭\ndiv 안의 div 클릭\n'

In [23]:
# 시장에게 바란다 접속 완료 

In [24]:

#Step 7.Beautiful Soup 로 본문 내용만 추출하기
from bs4 import BeautifulSoup
html_1 = driver.page_source
soup_1 = BeautifulSoup(html_1, 'html.parser')


In [25]:
html_1 = driver.page_source
soup_1 = BeautifulSoup(html_1, 'html.parser')

import math

total_cnt = soup_1.find('span','rplk_count').get_text()
total_cnt = int(total_cnt.replace('건',''))

print('총 %s 건의 민원신청이 검색되었습니다' % total_cnt)

cnt = int(input('이 중에서 몇 건을 수집하시겠습니까?: '))

page_cnt = math.ceil(cnt / 10)

print('%s 건의 데이터를 수집하기 위해 %s 페이지의 게시물을 조회합니다.' %(cnt,page_cnt))
print("\n")

총 340 건의 민원신청이 검색되었습니다
15 건의 데이터를 수집하기 위해 2 페이지의 게시물을 조회합니다.




rplk_count가 두 개 존재한다.

그래서 find()는 항상

첫 번째 것

만 가져온다.

지금은 우연히 맞지만 사이트 구조가 바뀌면 문제 발생 가능이 있다.

4️⃣ 조금 더 안정적인 코드

이렇게 작성하는 것이 좋다.

total_cnt = soup_1.find('span','rplk_count').get_text()

왜냐하면 이미

340건

이 첫 번째 span이기 때문이다.

숫자 처리 문제

현재 코드 결과는

340건

이다.

그런데 아래 코드에서 문제가 생긴다.

cnt = int(input('이 중에서 몇 건을 수집하시겠습니까?: '))

그리고 이후 계산

page_cnt = math.ceil(cnt / 10)

여기서 total_cnt가

340건

이라면 숫자가 아니라 문자열이다.

# 원래 내가 쓴 코드 

#Step 8. 총 검색 건수를 보여주고 수집할 건수 입력 받기
import math
total_cnt = soup_1.find('div','rp_lkment').find('span','rplk_count').get_text()
print('총 %s 건의 민원신청이 검색되었습니다' %(total_cnt))
cnt = int(input('이 중에서 몇 건을 수집하시겠습니까?: '))
page_cnt = math.ceil(cnt / 10)
print('%s 건의 데이터를 수집하기 위해 %s 페이지의 게시물을 조회합니다.' %(cnt,page_cnt))
print("\n")


여기까지는 잘 되는듯? 

In [26]:
titles = soup_1.find_all('a','rp_inlink')

print(len(titles))

10


In [27]:
for t in titles:
    print(t.get_text())

위례과천선 법조타운-북위례역(송파레이크파크호반써밋1,2차 사거리)-거여역 연장 강력 촉구
잠실 장미아파트 정비구역계획변경안의 재심의 상정을 간곡히 부탁드립니다.
6.5만 주민의 교통 생존권을 위한 우이신설선 남부 연장(옥수 직결) 건의
서대문구 영천동 모노레일 설치건에 대한 민원
서울 공급확대와 잠실 장미아파트 정비계획변경 심의 상정을 촉구드립니다
스포츠경기장
전기차 주차환경 개선 요청
산후조리경비 사용기간 연장 요청
5511번 차내혼잡 개선 및 증차 필요성
2026년 승용차 마일리지 제도 변경에 따른 민원 제기


In [28]:
title = soup_1.find('a','rp_inlink')

print(title.get_text())

위례과천선 법조타운-북위례역(송파레이크파크호반써밋1,2차 사거리)-거여역 연장 강력 촉구


In [29]:
titles = driver.find_elements(By.CLASS_NAME,'rp_inlink')

print(len(titles))

for t in titles:
    print(t.text)

10
위례과천선 법조타운-북위례역(송파레이크파크호반써밋1,2차 사거리)-거여역 연장 강력 촉구
잠실 장미아파트 정비구역계획변경안의 재심의 상정을 간곡히 부탁드립니다.
6.5만 주민의 교통 생존권을 위한 우이신설선 남부 연장(옥수 직결) 건의
서대문구 영천동 모노레일 설치건에 대한 민원
서울 공급확대와 잠실 장미아파트 정비계획변경 심의 상정을 촉구드립니다
스포츠경기장
전기차 주차환경 개선 요청
산후조리경비 사용기간 연장 요청
5511번 차내혼잡 개선 및 증차 필요성
2026년 승용차 마일리지 제도 변경에 따른 민원 제기


In [30]:
titles = driver.find_elements(By.CLASS_NAME,'rp_inlink')

print(titles[0].text)

titles[0].click()

위례과천선 법조타운-북위례역(송파레이크파크호반써밋1,2차 사거리)-거여역 연장 강력 촉구


In [31]:
html_1 = driver.page_source
soup_1 = BeautifulSoup(html_1, 'html.parser')

# 제목
title = soup_1.find('h3','sclinkage_title').get_text()

# 작성일
date = soup_1.find_all('span','sclinkage_evalue')[0].get_text()

# 내용
content = soup_1.find('span','rp_indata').get_text("\n")

print("제목:", title)
print("작성일:", date)
print("내용:", content[:100])  # 너무 길어서 앞부분만

제목: 위례과천선 법조타운-북위례역(송파레이크파크호반써밋1,2차 사거리)-거여역 연장 강력 촉구
작성일: 2026-02-25
내용: 안녕하세요.



저는 서울시 위례 송파레이크파크호반써밋1,2차 아파트 주민입니다.



2024년 11월 8일 국토교통부에서 발표한 위례과천선 민자적격성 조사 통과 보도참고자료에


In [32]:
#title = soup_1.find('h3', 'sclinkage_title').get_text()
#print(title)

In [33]:
#date = soup_1.find_all('span', 'sclinkage_evalue')[0].get_text()
#print(date)

In [34]:
#content = soup_1.find('span', 'rp_indata').get_text("\n")
#print(content)

#    1 페이지 반복
#      ↓
#    2 게시글 10개 찾기
#      ↓
#    3 게시글 클릭
#      ↓
#    4 제목 / 날짜 / 내용 수집
#      ↓
#    5 목록으로 돌아가기

In [35]:
html = driver.page_source
soup = BeautifulSoup(html,'html.parser')

post = soup.find('div','rp_inlink_item')

link = post.find('a','rp_inlink')['href']

print(link)

driver.execute_script(link)

AttributeError: 'NoneType' object has no attribute 'find'

In [ ]:
# Step 9. 데이터 수집하기
no2=[]
title2=[]
date2=[]
contents2=[]


no = 1

for a in range(1, page_cnt+1):

    print("\n")
    print("%s 페이지 내용 수집 시작 =======================" % a)

    html = driver.page_source
    soup = BeautifulSoup(html,'html.parser')

    # 게시글 목록
    content_list = soup.find_all('div','rp_inlink_item')

    for i in content_list:

        try:
            # 제목
            title = i.find('a','rp_inlink').get_text().strip()

            # javascript 링크
            link = i.find('a','rp_inlink')['href']

        except:
            continue

        print("\n")
        print("%s 번째 정보를 추출하고 있습니다============" % no)

        # 제목 저장
        title2.append(title)
        print("1. 제목:",title)

        # 게시글 클릭 (javascript 실행)
        driver.execute_script(link)
        time.sleep(2)

        # 상세페이지 HTML
        html_1 = driver.page_source
        soup_1 = BeautifulSoup(html_1,'html.parser')

        # 공개일
        try:
            date = soup_1.find_all('span','sclinkage_evalue')[0].get_text().strip()
        except:
            date = "공개일 없음"

        date2.append(date)
        print("2. 공개일:",date)

        # 상담내용
        try:
            cont = soup_1.find('span','rp_indata').get_text("\n").strip()
        except:
            cont = "내용 없음"

        contents2.append(cont)
        print("3. 상담내역:",cont)

        # 목록으로 돌아가기
        driver.back()
        time.sleep(2)

        no += 1

        if no > cnt:
            break

    if no > cnt:
        break

    # 다음 페이지 이동
    if no > cnt:
        break

# 다음 페이지 이동
    if a < page_cnt:
        driver.execute_script("move(%s)" % (a+1))
        time.sleep(2)
print("요청하신 작업이 완료되었습니다")



1 페이지 내용 수집 시작 =======================


1 번째 정보를 추출하고 있습니다============
1. 제목: 위례과천선 법조타운-북위례역(송파레이크파크호반써밋1,2차 사거리)-거여역 연장 강력 촉구
2. 공개일: 2026-02-25
3. 상담내역: 안녕하세요.



저는 서울시 위례 송파레이크파크호반써밋1,2차 아파트 주민입니다.



2024년 11월 8일 국토교통부에서 발표한 위례과천선 민자적격성 조사 통과 보도참고자료에 의하면 위례과천선 동쪽 구간은 송파구 법조타운으로 연결되는 노선으로 확정되었습니다. 위례 신도시의 교통불편을 해소하기 위해 계획되었던 위례과천선은 최초 노선안인 북위례지역을 경유하여 거여역으로 연결되는 노선이 아닌 송파구 법조타운으로 연결되는 노선으로 확정되었습니다. 

위례과천선은 위례 신도시의 교통불편을 해소하기 위해 계획되었고 위례 신도시 아파트 분양 납부금에는 교통 불모지였던 위례 신도시의 교통망을 확충하기 위한 교통분담금이 포함되어 있습니다. 따라서 위례를 거쳐가거나 위례와 이어지지 않는 위례과천선은 그 어떤 이유로도 납득할 수 없고 절대로 수용할 수 없습니다. 위례과천선 신설에 따른 수혜의 직접 대상이 되어야 하는 위례 신도시 주민들은 정작 교통망 확충의 혜택을 받지 못하게 되어 위례 신도시 주민으로서 너무나 참담하고 부당한 처사라고 생각합니다.

그러므로 위례과천선 동쪽 구간은 위례과천선 최초 원안을 반영하여 송파구 법조타운이 아니라 송파구 법조타운에서 북위례지역을 경유하여 거여역으로 연결되는 노선으로 확정되어야 합니다. 위례과천선 북위례지역 경유를 통한 거여역 연장은 한시도 지체할 수 없는 긴급하고 중대한 사안이며 다음과 같은 이유에서 즉시 확정되어야 합니다.



1. 북위례와 거여/마천 일대에 신축과 재건축 아파트가 속속 준공되고 입주가 완료되면서 거주 인구가 대폭 증가하고 그에 따라 교통수요는 폭발적으로 증가하였지만 교통망 확충은 한참 부족합니다. 북위례 지역의 A1-1(11단지) 국

In [ ]:
# gpt

#Step 9. 데이터 수집하기
no2=[]
title2=[]
date2=[]
contents2=[]

no = 1

for a in range(1,page_cnt+1):

    print("\n")
    print("%s 페이지 내용 수집 시작 =======================" %a)

    html = driver.page_source
    soup = BeautifulSoup(html, 'html.parser')

    content_list = soup.find_all('div','rp_inlink_item')

    for i in content_list:

        try:
            title = i.find('a','rp_inlink').get_text().strip()
            url = i.find('a','rp_inlink')['href']
        except:
            continue

        print("\n")
        print("%s 번째 정보를 추출하고 있습니다============" %no)

        # 제목 저장
        title2.append(title)
        print("1. 제목:",title)

        # 게시글 이동 (javascript 실행)
        driver.execute_script(url)
        time.sleep(2)

        html_1 = driver.page_source
        soup_1 = BeautifulSoup(html_1,'html.parser')

        # 공개일
        try:
            date = soup_1.find_all('span','sclinkage_evalue')[0].get_text().strip()
        except:
            date = "공개일 없음"

        date2.append(date)
        print("2. 공개일:",date)

        # 상담내용
        try:
            cont = soup_1.find('span','rp_indata').get_text("\n").strip()
        except:
            cont = "내용 없음"

        contents2.append(cont)
        print("3. 상담내역:",cont)

        # 목록으로 돌아가기
        driver.back()
        time.sleep(2)

        no += 1

        if no > cnt:
            break

    if no > cnt:
        break

    # 다음 페이지
    try:
        driver.find_element(By.LINK_TEXT,'다음').click()
        time.sleep(2)
    except:
        print("다음 페이지 없음")
        break

print("요청하신 작업이 완료되었습니다")



1 페이지 내용 수집 시작 =======================
다음 페이지 없음
요청하신 작업이 완료되었습니다


In [ ]:
#Step 9. 데이터 수집하기
no2=[] # 게시글 번호 컬럼
title2=[ ] # 게시글 제목 컬럼
date2=[ ] # 게시글 날짜 컬럼
contents2=[] # 상담내용
full_url2=[] # 논문 원본 URL

no = 1 # 게시글 번호 초기값

for a in range(1,page_cnt+1) :
    print("\n")
    print("%s 페이지 내용 수집 시작합니다 =======================" %a)

    html = driver.page_source
    soup = BeautifulSoup(html, 'html.parser')
    content_list = soup.find('div','rp_tb_wrap default_skin complaint_mayor_hope_list').find_all('div','rp_inlink_item')

    for i in content_list:
        # 여기서 바로 들어 가야겠네?? 아니면 제목 뽑고 들어가거나 ?? 

        # 민원신청 제목 체크하기
        try:
            title = soup_1.find('a','rp_inlink').get_text().strip()
        except :
            continue
        else :
            # 1.게시글 번호
            print("\n")
            print("%s 번째 정보를 추출하고 있습니다============" %no)
            no2.append(no)
            print("1.번호 : %s" %no)

            # 2. 민원신청 제목
            title2.append(title.strip())
            print("2.제목 : %s" %title.strip())

            # 3. 민원목록 안의 상세 내역에서 추출할 수 있음.
            url_1 = i.find('a`','rp_inlink_item').find('a')['href']
            full_url = 'https://eungdapso.seoul.go.kr/main.do'+url_1
            time.sleep(1)
            driver.get(full_url)

            html_1 = driver.page_source
            soup_1 = BeautifulSoup(html_1, 'html.parser')

            # 4. 공개일 
            try :
                date_1 =i.find('span','sclinkage_evalue').find_all('span')
                date_2 = date_1[1].get_text().strip()
            except :
                date_2='공개일가 없습니다'
                date2.append(date_2)
                print("4.공개일 : %s" %date_2)
            else :
                date2.append(date_2)
                print("4.공개일 : %s" %date_2)
            
            try :
                cont=soup_1.find("span","rp_indata").find('div','restd_incont').get_text().replace("\n","").strip()
            except :
                cont='상담내용이 없습니다'
                contents2.append(cont)
                print("5.상담내용 : %s" %cont)
            else :
                contents2.append(cont)
                print("5.상담내용 : %s" %cont)



            driver.back() # 이전 페이지로 돌아가기

            time.sleep(2)

            no += 1

            if no > cnt :
                break

    a += 1
    b = str(a)

    try :
        driver.find_element(By.LINK_TEXT ,'%s' %b).click()
    except :
        driver.find_element(By.LINK_TEXT,'다음 페이지로').click()

print("요청하신 작업이 모두 완료되었습니다")




1 페이지 내용 수집 시작합니다 =======================


2 페이지 내용 수집 시작합니다 =======================
요청하신 작업이 모두 완료되었습니다


In [ ]:
print(len(no2))
print(len(title2))
print(len(date2))
print(len(contents2))

0
0
0
0


In [ ]:

# Step 10. 수집된 데이터를 xlsx와 csv 형태로 저장하기
# 현재 날짜와 시간으로 폴더 만들고 파일 이름 설정하기
import os

n = time.localtime()
s = '%04d-%02d-%02d-%02d-%02d-%02d' %(n.tm_year, n.tm_mon, n.tm_mday, n.tm_hour, n.tm_min, n.tm_sec)

os.makedirs(f_dir+'RISS'+'-'+s+'-'+'학위논문')

fc_name = f_dir+'RISS'+'-'+s+'-'+'학위논문'+'\\'+'RISS'+'-'+s+'-'+'학위논문'+'.csv'
fx_name = f_dir+'RISS'+'-'+s+'-'+'학위논문'+'\\'+'RISS'+'-'+s+'-'+'학위논문'+'.xlsx'

# 데이터 프레임 생성 후 xlsx, csv 형식으로 저장하기
import pandas as pd

df = pd.DataFrame()
df['번호']=no2
df['제목']=pd.Series(title2)
df['날짜']=pd.Series(date2)
df['상담내용']=pd.Series(contents2)

# xls 형태로 저장하기
df.to_excel(fx_name,index=False)

# csv 형태로 저장하기
df.to_csv(fc_name,index=False, encoding="utf-8-sig")

print('요청하신 데이터 수집 작업이 정상적으로 완료되었습니다')

요청하신 데이터 수집 작업이 정상적으로 완료되었습니다


참고

In [ ]:
#spans = i.find('p','etc').find_all('span')

#author = spans[0].text
#company = spans[1].text
#date = spans[2].text
#degree = spans[3].text

In [ ]:
#1️⃣ 태그로 찾기
#find('div')
#2️⃣ class로 찾기
#find('span','writer')
#3️⃣ 여러개 찾고 인덱스로 구분
#find_all('span')[2]